README.md

```diff
@@ -1,277 +1,899 @@
+All users get a single swap address where all of their one-way swaps will be located. This universal
+validator can enforce all possible swaps for the user. And since users only need to worry about a
+single address, it keeps the UI simple.
 
-## Specification
-The protocol is composed of one validator script per swap pair, which is used to derive the minting script for the same pair. 
+##### Minting Redeemers
+
+Two minting redeemers are introduced here, their usage is explained further below.
+
+```Haskell
+data BeaconRedeemer
+  = CreateSwap 
+  | BurnBeacons
+```
 
-### Components
-An outline of the addresses, tokens, datums, and redeemers used in Cardano-Swaps.
+##### Validator Redeemers
 
-#### Swap Address
-Every swap-pair gets its own swap-address that is guarded by a pair-specific validator/spending script. This is accomplished with the help of the `SwapConfig` parameter:
+Two validator redeemers are introduced here, their usage is explained further below.
+
+```Haskell
+data SwapRedeemer
+  = CloseOrUpdate
+  | Swap
+```
+
+Only the owner (signified by the address' staking credential) can use the `CloseOrUpdate` redeemer.
+Anyone can use the `Swap` redeemer as long as the swap conditions are met.
+
+##### Datums
+Inline datums are used for all UTxOs. All UTxOs get the same type of datum:
 
 ``` Haskell
-data SwapConfig = SwapConfig
-  { swapOffer :: (CurrencySymbol,TokenName)
-  , swapAsk :: (CurrencySymbol,TokenName)
+data SwapDatum = SwapDatum
+  { beaconId :: CurrencySymbol -- ^ Beacon policy id for one-way swaps.
+  , pairBeacon :: TokenName -- ^ Trading pair beacon asset name for this swap.
+  , offerId :: CurrencySymbol -- ^ Offer policy id for this swap.
+  , offerName :: TokenName -- ^ Offer asset name for this swap.
+  , offerBeacon :: TokenName -- ^ Offer beacon asset name for this swap.
+  , askId :: CurrencySymbol -- ^ Ask policy id for this swap.
+  , askName :: TokenName -- ^ Ask asset name for this swap.
+  , swapPrice :: Rational -- ^ The desired swap ratio: Ask/Offer.
+  , prevInput :: Maybe TxOutRef -- ^ The output reference for the corresponding swap input.
   }
 ```
 
-**Each combination of `swapOffer` and `swapAsk` results in a different spending script. All spending scripts are identical except for this extra parameter. The hash of the resulting spending script is used as an extra parameter to the beacon minting policy. *All beacon minting policies are identical except for this extra parameter.* As a result, every `SwapConfig` will also have its own unique beacon policy.**
+The beacon information in the datum prevents misuse of beacons. The validator forces all beacons to
+be burned instead of being withdrawn. This ensures that beacons can only ever be located at
+swap addresses. **If the wrong beacon information is supplied, assets can be locked forever.** For
```

Comment:
That specific UTxO will be unspendable. This is the only consequence. This UTxO can never have a beacon token because the datum is misconfigured so other users will never see it when they query the DEX. I can change the wording to point this out.

Reply:
I think changing "assets" to "UTxO" should suffice:

> If the wrong beacon information is supplied, the UTxO can be locked forever.

Reply:
That specific wording change is helpful, but perhaps due to my own incomplete reading or understanding I don't see how a UTxO still in the user's possession (by retaining the private key for its spending) can become unspendable.  I feel I may be missing something really basic here, but if not then perhaps an explanation of how the UTxO gets "locked" (and what exactly this means) can be put into the document.

Reply:
It is a little technical. Basically, the spending script doesn't know what the beacon is by default. You need to tell the spending script which token is the beacon token by using the appropriate fields in the datum. The spending script then requires that those beacons *must be burned* in order to close the swap. Beacons cannot be withdrawn from a DEX address. What would happen if the datum is misused and says ada is the beacon? The spending script doesn't care, it will still demand that the asset set as the beacon in the datum must be burned instead of withdrawn. But ada is not burnable. Therefore, this requirement is unsatisfiable and thus, the UTxO is unspendable.

It is not possible to have beacons stored with misconfigured datums but it is possible to create UTxOs without beacons that do have misconfigured datums. Whether or not the misconfigured UTxO is spendable depends on whether or not the asset set as the beacon is burnable.

Reply:
I guess at a high level it can be summed up as: 

Beacons should not be allowed to be withdrawn from a DEX address; it should only be burned. But since the spending script must rely on the datum to know which asset is the beacon, misusing the datum can result in the spending script protecting the wrong asset.

Reply:
OK, then unless I've misunderstood, the part I was missing (and is also missing from the above elaboration) must be: the funds to be DEXed at that point are not in your regular assets, but are at UTxO's that you designate for Cardano-Swaps.  These particular addresses (not your regular ones) are governed by spending scripts that might no longer be satisfiable if the Beacon Tokens are used improperly.

If that is true then I think it would help to also include this general point, because I didn't readily understand the problem... and so might other technical but uninitiated readers.

Reply:
You are correct in saying that it is the swap UTxO that can be locked (ie, your actual limit order), not your personal funds. Although I don't think it is accurate to say "if the Beacon Tokens are used improperly" since the true beacon tokens can never be misused by design. This locking is only possible when the user doesn't use the true beacon tokens at all. To use the tagging analogy, locking is only possible if a user does not try to tag the Swap UTxO in a way to make it discoverable by others. The act of trying to properly tag it is enough to prevent this locking since the minting of the tag will fail if the datum is wrong and provide you with an error message explaining what is wrong.

Would saying "the Swap UTxO" instead of just "the UTxO" suffice? Explaining it makes it easy to get lost in the weeds and I think most developers can figure it out.

Reply:
I am sure most developers can deduce it... since I'm not a developer (more of an integrator, architect, project manager, etc.) I didn't get it right away.  I might have if I'd read the document a few more times but we can't always rely upon that.

There will be other non-developers reading this page mainly to understand the value of implementations, so in this case (and maybe others) it does need to be broadened a bit beyond the audience of seasoned developers to allow all interested parties to immediately understand & qualify the risk. 😇

Reply:
I apologize if my comment was slightly offensive, that was not my intention. How's this?

> The beacon information in the datum prevents misuse of beacons. The validator forces all beacons to be burned instead of being withdrawn. This ensures that beacons can only ever be located at swap addresses. **If the wrong beacon information is supplied to the datum, the resulting Swap UTxO can be locked forever.** Simply trying to store the proper beacons in the Swap UTxO is enough to prevent accidental locking due to misconfigured datums. All minting policies check that the datums have the proper information; the mint transaction will fail with the relevant error message if they do not.

Reply:
no worries on my account; I didn't take it that way 😎 ... and for the general community I think the explanation above will work well to address that question in advance:
- Less informed readers wouldn't be concerned by the boldface sentence, since it implies that someone would already have to be using the method very deliberately in order for funds to be potentially locked & lost.
- Most of those less technical readers will further conclude that some implementation of Cardano-Swaps will be handling those details & therefore general users wouldn't be at risk by executing the instructions incorrectly themselves. 🤔

Reply:
> If the wrong beacon information is supplied, assets can be locked forever.
Maybe I am not understanding, but won't the user still be able to withdraw his locked assets from the swap using his staking key?

Reply:
No, if the datum is misconfigured, it is possible for the UTxO to be unspendable. The spending script doesn't actually know which asset is the beacon; it relies on the datum to tell it which one is the beacon. It then requires that asset to be burned instead of withdrawn whenever a swap is closed. What if the datum sways ada is the beacon?

Reply:
It does sound like you have thought about this problem deeply such that your design stops it from happening except if people deliberately act against the design.

However, I am still a little unsure if it is necessary to do the extra checking that the beacon and datum align correctly in the case where the user is withdrawing using his staking key.  If the user can sign the staking key then why not allow the asset(s) to be withdrawn and the beacon gets deleted, without doing the check that the beacon is correct?  That way if the user somehow creates a swp with asset and misconfigured beacon, deliberately or otherwise, then he can still withdraw using the staking key and the misconfigured beacon still gets deleted.

If it is possible to create misconfigured beacons maliciously then this will still require the minimum Ada in the eUTxO, but wouldn't this be better if the user had a way of cleaning them up and getting the 2 Ada min deposit back?  Would such malicious beacons pollute the chain and make querying for valid swaps less efficient?

Reply:
Considering how difficult it is to misconfigure a datum, I don't want to add the logic for the ada exception which would hurt the composition for all users. I don't think it is that important to hurt performance.